lib/SILGen/SILGenExpr.cpp

```diff
@@ -3553,27 +3553,60 @@ RValue RValueEmitter::visitKeyPathExpr(KeyPathExpr *E, SGFContext C) {
   
   auto baseTy = rootTy;
   SmallVector<SILValue, 4> operands;
-  
-  auto lowerSubscriptOperands =
-    [this, &operands, E](const KeyPathExpr::Component &component) {
-      if (!component.getIndexExpr())
-        return;
-      
-      // Evaluate the index arguments.
-      SmallVector<RValue, 2> indexValues;
-      auto indexResult = visit(component.getIndexExpr(), SGFContext());
-      if (isa<TupleType>(indexResult.getType())) {
-        std::move(indexResult).extractElements(indexValues);
-      } else {
+
+  auto lowerSubscriptOperands = [this, &operands,
+                                 E](const KeyPathExpr::Component &component) {
+    if (!component.getIndexExpr())
+      return;
+
+    auto decl = dyn_cast<SubscriptDecl>(component.getDeclRef().getDecl());
+    assert(decl &&
+           "lowerSubscriptOperands must be called with a subscript decl");
+
+    // Evaluate the index arguments.
+    SmallVector<RValue, 2> indexValues;
+    RValue indexResult;
+    if (auto paren = dyn_cast<ParenExpr>(component.getIndexExpr())) {
+      auto param = decl->getIndices()->get(0);
+      if (param->isDefaultArgument())
+        indexResult = visit(param->getDefaultValue(), SGFContext());
```

Comment:
Unfortunately, I believe this will run into the same problem as before with cross-module default arguments, as we don't deserialize an expression for them. Instead of trying to directly emit SIL for the expression, I think we should be trying to emit an apply of the appropriate default argument generator (see `DelayedArgument::emitDefaultArgument`). The benefit of using `ArgEmitter` is that it should handle this case for us, did you have any luck with that?

Reply:
I tried `SGF.SGM.emitDefaultArgGenerators` which tried to re-emit the subscript function. Then I tried to use `emitGeneratorFunction` directly which got further before having issues. Neither ended up working. I'll keep playing with it, though. 

Also, `emitDefaultArgGenerators` uses `param->getDefaultValue` which means it would have the same issue, right? Or is it OK because it calls `emitOrDelayFunction`?

Reply:
Actually, looking at [the code](https://github.com/apple/swift/blob/master/lib/SILGen/SILGen.cpp#L1062-L1073) again it looks like the call to `getDefaultValue` _isn't_ in the lambda which means it would be called around the same time either way.

Reply:
@zoecarver `emitDefaultArgGenerators` works because it's called when emitting the function declaration itself, and therefore has access to the default argument (e.g for a function with a default arg in Module A that gets used from Module B, `emitDefaultArgGenerators` only gets called when emitting Module A).

You shouldn't need to call it (or `emitGeneratorFunction`) when emitting default arguments at a call-site, as the appropriate default argument generator should have already been emitted. What we want to emit is an apply of that generator function (which is what `DelayedArgument::emitDefaultArgument` does).

Reply:
That makes sense, thanks for explaining. `DelayedArgument` only exists in `SILGenApply.cpp` and (AFAICT) it doesn't touch key paths. How should I access `DelayedArgument::emitDefaultArgument`?

Reply:
@zoecarver You'll probably want to break off a function that deals with key path subscript argument lowering into SILGenApply.cpp so it has access to those functions. If you're able to use `ArgEmitter`, then it'll give you a list of delayed arguments which you can emit by calling `emitDelayedArguments`, which will call through to `DelayedArgument::emitDefaultArgument` for you.

Reply:
I added a function in SILGenApply.cpp where I call `emitSingleArg` via `ArgEmitter` for every element in an array of collected index expressions. But, it still errors when it receives a `DefaultArgumentExpr`. Any ideas (I can push my code if that would be helpful)?

Also, I noticed that `visitSubscriptDecl` calls `emitDefaultArgGenerators`. Maybe that should be updated to use `ArgEmitter`  as well. I'll see if I can test it with multiple modules and see if there's an issue. 

Thanks for all the help with this :)

Reply:
@zoecarver I think you may want to use `PreparedArguments` in order to decompose the tuple/paren expr for you, which you can then pass to `ArgEmitter::emitPreparedArgs`. Actually I've just noticed we have `SILGenFunction::prepareSubscriptIndices` which is used to decompose arguments for regular subscripts, I wonder if we can re-use that here?

Regarding `visitSubscriptDecl`, note that this is for the subscript declaration, not the call-site. `SILGenLValue::visitSubscriptExpr` is where you want to look for regular subscript emission. The argument emission logic is a bit convoluted, but it'll eventually wind up calling various routines depending on the exact accessor being used, e.g `SILGenFunction::emitGetAccessor`. This logic uses `CallEmission` which will use `ArgEmitter` for the arguments (I don't believe we can use `CallEmission` ourselves as we don't want an actual call, we just want the arguments).

Reply:
Updated to use `ArgEmitter`! This definitely feels like the right way to do it (and it makes all argument parsing more general).